### PR TITLE
Adding SDK for Microsoft.DataProtection

### PR DESCRIPTION
**Pull request by mayankagg9722**
# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/master/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

**Comment by mayankagg9722** [comment]
I am facing issues with the deserialization of the request body with the correct "objectType" mentioned in the json.

**Comment by markcowl** [comment]
@mayankagg9722 Please see my response on your filed issue - you need to use the serialization/deserialization settings from the client to get polymorphic behavior on deserialization.

**Comment by mayankagg9722** [comment]
> @mayankagg9722 Please see my response on your filed issue - you need to use the serialization/deserialization settings from the client to get polymorphic behavior on serialization/deserialization.

Yeah, I did the and it worked for that issue.

@markcowl can you also please let me know about other error I am facing mentioned below? Though it is there in the Session record json.

**Error:**
`
System.InvalidOperationException: 'Queue empty for request GET /subscriptions/f2edfd5d-5496-4683-b94f-b3588c579009/providers/Microsoft.DataProtection/locations/southeastasia/operationStatus/MGNiY2U3OWQtYzU4Yi00OWI0LWIwY2QtMTEyNDBhZTk4Nzk3O2RlYTdhNzZmLTlhZTItNDkzZS04ZmI0LTZlNzQyYWM3ZDM0Yw==?api-version=2021-01-01'
`

**Session record json: (Line number: 800)**
`{
      "RequestUri": "/subscriptions/f2edfd5d-5496-4683-b94f-b3588c579009/providers/Microsoft.DataProtection/locations/southeastasia/operationStatus/MGNiY2U3OWQtYzU4Yi00OWI0LWIwY2QtMTEyNDBhZTk4Nzk3O2RlYTdhNzZmLTlhZTItNDkzZS04ZmI0LTZlNzQyYWM3ZDM0Yw==?api-version=2021-01-01",
      "EncodedRequestUri": "L3N1YnNjcmlwdGlvbnMvZjJlZGZkNWQtNTQ5Ni00NjgzLWI5NGYtYjM1ODhjNTc5MDA5L3Byb3ZpZGVycy9NaWNyb3NvZnQuRGF0YVByb3RlY3Rpb24vbG9jYXRpb25zL3NvdXRoZWFzdGFzaWEvb3BlcmF0aW9uU3RhdHVzL01HTmlZMlUzT1dRdFl6VTRZaTAwT1dJMExXSXdZMlF0TVRFeU5EQmhaVGs0TnprM08yUmxZVGRoTnpabUxUbGhaVEl0TkRrelpTMDRabUkwTFRabE56UXlZV00zWkRNMFl3PT0/YXBpLXZlcnNpb249MjAyMS0wMS0wMQ==",
      "RequestMethod": "GET",
      "RequestBody": "",
      "RequestHeaders": {
        "x-ms-client-request-id": [
          "1f015d8b-793d-463b-9997-32b8c464b97b"
        ],
        "Accept-Language": [
          "en-US"
        ],
        "User-Agent": [
          "FxVersion/4.6.29321.03",
          "OSName/Windows",
          "OSVersion/Microsoft.Windows.10.0.19042."
        ],
        "Content-Type": [
          "application/json; charset=utf-8"
        ],
        "Content-Length": [
          "98"
        ]
      },
      "ResponseHeaders": {
        "Cache-Control": [
          "no-cache"
        ],
        "Pragma": [
          "no-cache"
        ],
        "X-Content-Type-Options": [
          "nosniff"
        ],
        "x-ms-request-id": [
          "913287bd-7d1a-4ed8-bdad-d0d2a9432c86"
        ],
        "x-ms-client-request-id": [
          "1f015d8b-793d-463b-9997-32b8c464b97b"
        ],
        "Strict-Transport-Security": [
          "max-age=31536000; includeSubDomains"
        ],
        "Server": [
          "Microsoft-IIS/10.0"
        ],
        "x-ms-ratelimit-remaining-subscription-resource-requests": [
          "209"
        ],
        "x-ms-correlation-request-id": [
          "873ce447-d705-43e0-a250-c245a146c480"
        ],
        "x-ms-routing-request-id": [
          "SOUTHINDIA:20201208T135053Z:873ce437-d705-43e0-a250-c245a146c280"
        ],
        "Date": [
          "Tue, 08 Dec 2020 13:50:52 GMT"
        ],
        "Content-Length": [
          "484"
        ],
        "Content-Type": [
          "application/json"
        ],
        "Expires": [
          "-1"
        ]
      },
      "ResponseBody": "{\"id\":\"\/subscriptions\/f2edfd5d-5496-4683-b94f-b3588c579009\/providers\/Microsoft.DataProtection\/locations\/southeastasia\/operationStatus\/MGNiY2U3OWQtYzU4Yi00OWI0LWIwY2QtMTEyNDBhZTk4Nzk3OzI2OGIyZmMzLTExYzAtNDlkOC1iZTI4LTkyYjhkYjE5MGMyMA==\",\"name\":\"MGNiY2U3OWQtYzU4Yi00OWI0LWIwY2QtMTEyNDBhZTk4Nzk3OzI2OGIyZmMzLTExYzAtNDlkOC1iZTI4LTkyYjhkYjE5MGMyMA==\",\"status\":\"Succeded\",\"startTime\":\"2021-04-12T11:52:46.9915389Z\",\"endTime\":\"2021-04-12T11:53:08Z\"}",
      "StatusCode": 200
    }`

image
image

**Comment by markcowl** [comment]
@mayankagg9722 This error is from running a test that is using the test framework in playback mode, where previously recorded service responses are returned in response to api calls.  The error means that it could not find a response in the recording matching the request - think of the recording as a dictionary of queues of responses, where the key is the uri + verb of the request.  If there is no entry for a particular request (say, because the api-version doesn't match the api-version in the recording), or if the queue or responses is exhausted, (say, because the calling pattern in the test has changed), then you will get this message.

Generally, this means you need to re-record the test by running it in record mode.  Note that, in a serialization test in which you provide the requests and responses in the test code, you probably don't want to be using the test framework's recording mechanisms.

**Comment by mayankagg9722** [comment]
> Generally, this means you need to re-record the test by running it in record mode

- Can you please elaborate on "queue or responses is exhausted"?

- We have already mocked responses for azure-async headers but we are not able to run the tests completely as the mocked request and responses are not being used and it gave queue empty error. Can you please tell me what we can do here to fix this?

- This is the first time we are integrating the tests and nothing is changes in SDK and we recently generated this using "generate.ps1". And written few tests for this, but "CreateOrUpdateAsync" calls are failing because of "GetPutOrPatchOperationResultAsync".

Example: response result is coming from the mocked PUT call but the response for "GetPutOrPatchOperationResultAsync" is failing even after the call is mocked in the session record json.

**Comment by markcowl** [comment]
@mayankagg9722 How have you mocked the responses?  Mocking the responses would mean that you have customized the client or the http pipeline to return your synthetic json, how have you done this?

'The queue is exhausted' means that, every time during the test, when a call is made to a particular uri+verb, one of the responses from the queue of responses for that key is used up.  If there are more requests to that uri+verb than exist in the recording, then there is nothing to return, and you would get a message about the queue being empty.

It seems that you are trying to mix whatever mocking mechanism you are using for these synthesized json records with the standard test framework.  That is not likely to work, and not sure why you wouldn't just use the existing test framework for live recorded tests with the service, and test the serializers separately, in tests that do not use the live/recording test framework.

**Comment by mayankagg9722** [comment]
> @mayankagg9722 How have you mocked the responses? Mocking the responses would mean that you have customized the client or the http pipeline to return your synthetic json, how have you done this?
> 
> 'The queue is exhausted' means that, every time during the test, when a call is made to a particular uri+verb, one of the responses from the queue of responses for that key is used up. If there are more requests to that uri+verb than exist in the recording, then there is nothing to return, and you would get a message about the queue being empty.
> 
> It seems that you are trying to mix whatever mocking mechanism you are using for these synthesized json records with the standard test framework. That is not likely to work, and not sure why you wouldn't just use the existing test framework for live recorded tests with the service, and test the serializers separately, in tests that do not use the live/recording test framework.

@markcowl  Can you please help me with some documentation for the test recording so that session record json will be genereted in automated way and I don't need to manually put the entries there?

**Comment by mayankagg9722** [comment]
@markcowl I have recorded the json and was able to create live request and responses. This issue is now resolved. Can you please review the Pull request for the integration of NET SDK for Microsoft.DataProtection

File: sdk/dataprotection/Microsoft.Azure.Management.DataProtection/Microsoft.Azure.Management.DataProtection.sln [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+		{004937D0-0E95-47FA-A9D9-949D6224B79B}.Release|x86.ActiveCfg = Release|Any CPU
+		{004937D0-0E95-47FA-A9D9-949D6224B79B}.Release|x86.Build.0 = Release|Any CPU
+	EndGlobalSection
+	GlobalSection(SolutionProperties) = preSolution
+		HideSolutionNode = FALSE
+	EndGlobalSection
+	GlobalSection(ExtensibilityGlobals) = postSolution
+		SolutionGuid = {7BCCAB5F-5599-4596-A0CC-4E49ECD39DEF}
+	EndGlobalSection
+EndGlobal
```
Comment (markcowl) [inline sdk/dataprotection/Microsoft.Azure.Management.DataProtection/Microsoft.Azure.Management.DataProtection.sln:51]: remove 'upgraxdelog.htm' file below

File: sdk/dataprotection/Microsoft.Azure.Management.DataProtection/src/package-lock.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (markcowl) [inline sdk/dataprotection/Microsoft.Azure.Management.DataProtection/src/package-lock.json:1]: remove this file

File: sdk/dataprotection/Microsoft.Azure.Management.DataProtection/src/Microsoft.Azure.Management.DataProtection.Backup.csproj [hunk 3]
```diff
@@ -0,0 +1,27 @@
+<Project Sdk="Microsoft.NET.Sdk">
+	<!-- Please do not move/edit code below this line -->
+	<Import Project="$(RepoEngPath)/mgmt/AzSdk.reference.props" />
+	<!-- Please do not move/edit code below this line -->
+
+	<PropertyGroup>
+		<PackageId>Microsoft.Azure.Management.DataProtection.Backup</PackageId>
+		<Description>Provides developers with libraries for data protection under Azure Resource manager.</Description>
+		<Version>4.1.4-preview</Version>
```
Comment (markcowl) [inline sdk/dataprotection/Microsoft.Azure.Management.DataProtection/src/Microsoft.Azure.Management.DataProtection.Backup.csproj:9]: If this is the first time this package is being added, why is the version 4.1.4-preview?

The version should be 1.0.0-beta.1

File: sdk/dataprotection/Microsoft.Azure.Management.DataProtection/src/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Reflection;
+using System.Resources;
+
+[assembly: AssemblyTitle("Microsoft Azure DataProtection Library")]
+[assembly: AssemblyDescription("Provides management functionality for Microsoft Azure DataProtection Resources.")]
+
+[assembly: AssemblyVersion("4.0.0.0")]
```
Comment (markcowl) [inline sdk/dataprotection/Microsoft.Azure.Management.DataProtection/src/Properties/AssemblyInfo.cs:10]: same comment

**Review by markcowl — CHANGES_REQUESTED** [review]
Also you will need to add an entry for your project here to have your tests run in ci: https://github.com/Azure/azure-sdk-for-net/blob/master/eng/pipelines/mgmt.yml#L67

**Comment by markcowl** [comment]
@mayankagg9722 Updated my review

File: sdk/dataprotection/Microsoft.Azure.Management.DataProtection/Microsoft.Azure.Management.DataProtection.sln (hunk 1 above)
Comment (mayankagg9722) [reply sdk/dataprotection/Microsoft.Azure.Management.DataProtection/Microsoft.Azure.Management.DataProtection.sln:51]: removed and added package in pipeline yaml

File: sdk/dataprotection/Microsoft.Azure.Management.DataProtection/src/package-lock.json (hunk 2 above)
Comment (mayankagg9722) [reply sdk/dataprotection/Microsoft.Azure.Management.DataProtection/src/package-lock.json]: removed this

File: sdk/dataprotection/Microsoft.Azure.Management.DataProtection/src/Microsoft.Azure.Management.DataProtection.Backup.csproj (hunk 3 above)
Comment (mayankagg9722) [reply sdk/dataprotection/Microsoft.Azure.Management.DataProtection/src/Microsoft.Azure.Management.DataProtection.Backup.csproj]: Changed the version

File: sdk/dataprotection/Microsoft.Azure.Management.DataProtection/src/Properties/AssemblyInfo.cs (hunk 4 above)
Comment (mayankagg9722) [reply sdk/dataprotection/Microsoft.Azure.Management.DataProtection/src/Properties/AssemblyInfo.cs]: changed version

**Comment by mayankagg9722** [comment]
> @mayankagg9722 Updated my review

@markcowl Resolved all the PR comments, please check once.

**Review by markcowl — APPROVED** [review]

**Pull request merged**